File: gcpc2024/jigsawpresent/submissions/accepted/erik.py
N, M, C = 28, 13, 4100
vis = [{} for i in range((2 * C + 1) * M)]
a = []

def finish(*args):
	print("no")
	for msk in args:
		print(bin(msk).count("1"), end = " ")
		for i in range(N):
			if (msk >> i) % 2 == 1:
				print(i + 1, end = " ")
		print()
	exit(0)

def dfs(x, y, m, f, mskL, mskR):
	if x == 0 and y == 0: finish(mskL, mskR)
	if m < 0 and f == 0:
		vis[x + M * C][y] = (mskL, mskR)
		return
	if m < M and f == 1:
		for s in [-1, 1]:
			i = s * x + M * C
			if 0 <= i < (2 * C + 1) * M and s * y in vis[i]:
				u, v = vis[i][s * y]
				if s == -1:
					u, v = v, u
				finish(mskL | v, mskR | u)
		return
	dfs(x, y, m - 1, f, mskL, mskR)
	dfs(x + a[m][0], y + a[m][1], m - 1, f, mskL | (2**m), mskR)
	dfs(x - a[m][0], y - a[m][1], m - 1, f, mskL, mskR | (2**m))

n = min(int(input()), N)
for i in range(n):
	y, x = map(int, input().split(' '))
	a += [(x, y)]
	dfs(x, y, i - 1, i >= M, 2**i, 0)
print("yes")

File: gcpc2024/jigsawpresent/submissions/accepted/erik.java
import java.util.*;
public class erik {
    static int N = 28, M = 13, C = 4100;
    static int[][] a = new int[N][2];
    static HashMap<Integer, int[]> vis[] = new HashMap[(2 * C + 1) * M];
    static void finish(int x, int y) {
        System.out.println("no");
        for (int msk: new int[]{x, y}) {
            System.out.print(Integer.bitCount(msk));
            for (int i = 0; i < N; i++)
                if ((msk >> i & 1) == 1)
                    System.out.print(" " + (i + 1));
            System.out.println();
        }
        System.exit(0);
    }
    static void dfs(int x, int y, int m, boolean f, int mskL, int mskR) {
        if (x == 0 && y == 0) finish(mskL, mskR);
        if (m < 0 && !f) {
            vis[x + M * C].put(y, new int[]{mskL, mskR});
            return;
        }
        if (m < M && f) {
            for (int s: new int[]{-1, 1}) {
                int i = s * x + M * C;
                if (0 <= i && i < (2 * C + 1) * M) {
                    int[] tmp = vis[i].get(s * y);
                    if (tmp == null)
                        continue;
                    if (s == -1)
                        tmp = new int[]{tmp[1], tmp[0]};
                    finish(mskL | tmp[1], mskR | tmp[0]);
                }
            }
            return;
        }
        dfs(x, y, m - 1, f, mskL, mskR);
        dfs(x + a[m][0], y + a[m][1], m - 1, f, mskL | (1 << m), mskR);
        dfs(x - a[m][0], y - a[m][1], m - 1, f, mskL, mskR | (1 << m));
    }
    public static void main(String[] args) {
        for (int i = 0; i < (2 * C + 1) * M; i++)
            vis[i] = new HashMap<Integer, int[]>();
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        n = Math.min(n, N);
        for (int i = 0; i < n; i++) {
            a[i][1] = in.nextInt();
            a[i][0] = in.nextInt();
            dfs(a[i][0], a[i][1], i - 1, i >= M, 1 << i, 0);
        }
        System.out.println("yes");
    }
}
